### ISSUE TITLE
Paste issue in IPython 7.0.1

### ISSUE THREAD

**Issue by gruevyhat** [jpalardy/vim-slime#172]
I recently upgraded IPython to 7.0.1, and for some reason vim-slime has stopped working. When I hit `<ctrl-c><ctrl-c>` in Vim, it successfully inserts `%cpaste` into IPython, but then just stops---it doesn't paste any of the actual python commands. If I hit `<ctrl-c>` within IPython, it suddenly pastes all the lines, but only one line at a time, not as a code block; finally it pastes `--`, prompting a syntax error. I checked to make sure .slime_paste contains the python content, and it does. I have `let g:slime_python_ipython = 1` set in .vimrc. I get the same behavior in Gnu screen and tmux. Any help would be greatly appreciated.

**Comment by jpalardy** [comment]
Hi @gruevyhat 

You might want to look at the other [iPython issues](https://github.com/jpalardy/vim-slime/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+ipython) and [this PR](https://github.com/jpalardy/vim-slime/pull/170).

Also:
* what's your setup?
* did you try a barebone vim with only vim-slime?

**Comment by jpalardy** [comment]
Damn.

Who wants to take a stab at this?

**Comment by gmnimrod** [comment]
I am having the same problem.
My setup: IPython 7.2.0, neovim 0.2.2, tmux 2.6, vim-slime 06525ceee2457bed7c7718fef0b4fe715651c392.

I have a possible explanation.
The text that Slime prepares is copied to ipython before the %cpaste command is executed.
When %cpaste command is executed, ipython enters the special cpaste mode.
Slime already copied all the text, so ipython just hangs, waiting for input.
After breaking with ctrl-c, ipython resumes reading lines from the regular input stream.
The code is already there. Ipython executes the code line by line.

Another observation: When using slime to send a large number of lines (~1000 on my setup) the second half of the code is pasted in cpaste mode, and the first half is pasted after exiting cpaste mode. This behavior is consistent with the above explanation.

This issue seems related to #142.

Adding a delay in `slime#send`, as suggested by @fx-kirin in #142 solves this issue.
I think that it is enough to add the delay once, only after "%cpaste -q\n".
So, the solution might be something like this:

```vim
function! slime#send(text)
  call s:SlimeGetConfig()

  let pieces = s:_EscapeText(a:text)
  " begin change
  if exists('g:slime_python_ipython')
    call s:SlimeDispatch('Send', b:slime_config, pieces[0])
    sleep 100m
    let pieces = pieces[1:]
  end
  " end change
  for piece in pieces
      call s:SlimeDispatch('Send', b:slime_config, piece)
  endfor
endfunction
```

Python specific code should go in ftplugin/python/slime.vim.
The condition on `g:slime_python_ipython` could be dropped, but that would change the behavior for all use cases, not just for ipython 5+.
I suggest adding a special non-string marker in `pieces`, which will signal `slime#send` to pause.

I suggest something along these lines:

```vim
" in autoload/slime.vim

function! slime#send(text)
  call s:SlimeGetConfig()

  let pieces = s:_EscapeText(a:text)
  for piece in pieces
    " begin change
    if piece == v:null
      execute 'sleep' g:slime_dispatch_pause . 'm'
    else
      call s:SlimeDispatch('Send', b:slime_config, piece)
    end
    " end change
  endfor
endfunction

" in ftplugin/python/slime.vim

if !exists("g:slime_dispatch_pause")
  let g:slime_dispatch_pause = 100
end

function! _EscapeText_python(text)
  if exists('g:slime_python_ipython') && len(split(a:text,"\n")) > 1
    " begin change
    return ["%cpaste -q\n", v:null, a:text, "--\n"]
    " end change
  else
    let empty_lines_pat = '\(^\|\n\)\zs\(\s*\n\+\)\+'
    let no_empty_lines = substitute(a:text, empty_lines_pat, "", "g")
    let dedent_pat = '\(^\|\n\)\zs'.matchstr(no_empty_lines, '^\s*')
    let dedented_lines = substitute(no_empty_lines, dedent_pat, "", "g")
    let except_pat = '\(elif\|else\|except\|finally\)\@!'
    let add_eol_pat = '\n\s[^\n]\+\n\zs\ze\('.except_pat.'\S\|$\)'
    return substitute(dedented_lines, add_eol_pat, "\n", "g")
  end
endfunction
```

A global variable is used for configuring the length of the delay, as suggested by @jpalardy in #142.

**Comment by jpalardy** [comment]
@gmnimrod great explanation

would you like to send me a PR for this?

**Comment by gmnimrod** [comment]
I would be glad to contribute :smile: 
I submitted a pull request: #186.

**Comment by oblitum** [comment]
I have paste issues with [pgcli](https://github.com/dbcli/pgcli) :-/ Any help appreciated. I get extra tabulation on each new line, when pasted.

**Comment by jpalardy** [comment]
Hi @oblitum 

Could you please open another issue for this? Write up what you can and we'll see what we can do to help.

**Comment by oblitum** [comment]
@jpalardy thanks. I've created #191.
